sql-plugin/src/main/java/com/nvidia/spark/rapids/JCudfUtil.java

```diff
@@ -0,0 +1,582 @@
+        varSizeColIndex = ind;
+      }
+      res = alignOffset(byteCursor, JCUDF_VAR_LENGTH_FIELD_ALIGNMENT);
+      byteCursor = res + JCUDF_VAR_LENGTH_FIELD_SIZE;
+      return res;
+    }
+
+    private void resetMeta() {
+      varSizeColIndex = attributes.length;
+      byteCursor = 0;
+      estimateRowSize = -1;
+      validityBytesOffset = 0;
+    }
+
+    private void setValidityBits(int fieldsCount) {
+      validityBytesOffset = byteCursor;
+      byteCursor += calculateBitSetWidthInBytes(fieldsCount);
+    }
+
+    /**
+     * This is used internally to estimate the memory needed by the row.
+     * The value calculated is 8-bytes aligned.
+     * It loops on variable-width columns using a moving-cursor that represents the offset
+     * of the data.
+     * Note that the calculation requires that {@link #initFixedWidthSection(int[])} has
+     * been called at some point before. Otherwise, the {@link #byteCursor} will be 0.
+     */
+    private void calculateEstimateSize() {
+      // at this point the cursor points to the beginning of the variableWidth section
+      estimateRowSize = byteCursor;
+      if (hasVarSizeData()) {
+        for (int ind = varSizeColIndex; ind < attributes.length; ind++) {
+          DataType dataType = attributes[ind].dataType();
+          if (isFixedLength(dataType)) {
+            continue;
+          }
+          DType rapidsType = GpuColumnVector.getNonNestedRapidsType(dataType);
+          estimateRowSize = advanceDataCursorAndGet(estimateRowSize, rapidsType);
+        }
+      }
+      // for row estimate, we need just to align to row alignment
+      estimateRowSize = alignOffset(estimateRowSize, JCUDF_ROW_ALIGNMENT);
+    }
+
+    /**
+     * Increments the data cursor by the length of the column type.
+     * It takes into considerations the datatype alignment.
+     * See {@link #getDataAlignmentForDataType(DType)}.
+     *
+     * @param dataOffsetCursor current offset of the data values.
+     * @param rapidsType the type of the current column.
+     * @return data cursor (1-byte aligned) after advancing by the length of the data.
+     */
+    private int advanceDataCursorAndGet(int dataOffsetCursor, DType rapidsType) {
+      // For variable-length, alignment and length are not equivalent.
+      int length = getEstimateSizeForVarLengthTypes(rapidsType);
+      int alignment = getDataAlignmentForDataType(rapidsType);
+      dataOffsetCursor = alignOffset(dataOffsetCursor, alignment);
+      return dataOffsetCursor + length;
+    }
+
+    /**
+     * Used to check whether a row is variable-width or not.
+     * This assumes that an empty schema is "fixed-width".
+     *
+     * @return true if the schema is non-empty and at least one column is variable-width size.
+     */
+    public boolean hasVarSizeData() {
+      return attributes.length > 0 && varSizeColIndex != attributes.length;
+    }
+
+    public int getValidityBytesOffset() {
```

Comment:
I'm noticing some new functions without comments. I'm not sure the comment on function policy of the java code though, so I'm just going to comment that I see it. :)